src/math/fft/fft.c

```diff
@@ -0,0 +1,228 @@
+ * \param[in] ifft - set to 1 for IFFT and 0 for FFT.
+ * \param[out] outb - pointer to output buffer.
+ */
+static void fft_internal(struct icomplex32 *inb, struct icomplex32 *outb,
+			 uint32_t size, bool ifft)
+{
+	struct icomplex32 tmp1;
+	struct icomplex32 tmp2;
+	int *bit_reverse_idx;
+	int depth;
+	int lim = 1;
+	int len = 0;
+	int top;
+	int bottom;
+	int index;
+	int i;
+	int j;
+	int k;
+	int m;
+	int n;
+
+	/* calculate the exponent of 2 */
+	while (lim < size) {
+		lim <<= 1;
+		len++;
+	}
+
+	/* convert to complex conjugate for ifft */
+	if (ifft) {
+		for (i = 0; i < size; i++)
+			icomplex_conj(&inb[i]);
+	}
+
+	bit_reverse_idx = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				  lim * sizeof(int32_t));
+
+	/* step 1: re-arrange input in bit reverse order, and shrink the level to avoid overflow */
+	for (i = 1; i < lim; ++i) {
+		bit_reverse_idx[i] = (bit_reverse_idx[i >> 1] >> 1) |
+				     ((i & 1) << (len - 1));
+		icomplex_shift(&inb[i], -len, &outb[bit_reverse_idx[i]]);
+	}
+
+	rfree(bit_reverse_idx);
+
+	/* step 2: loop to do FFT transform in smaller size */
+	for (depth = 1; depth <= len; ++depth) {
+		m = 1 << depth;
+		n = m >> 1;
+		i = FFT_SIZE_MAX / m;
+
+		/* doing FFT transforms in size m */
+		for (k = 0; k < lim; k += m) {
+			/* doing one FFT transform for size m */
+			for (j = 0; j < n; ++j) {
+				index = i * j;
+				top = k + j;
+				bottom = top + n;
+				tmp1.real = twiddle_real[index];
+				tmp1.imag = twiddle_imag[index];
+				/* calculate the accumulator: twiddle * bottom */
+				icomplex32_mul(&tmp1, &outb[bottom], &tmp2);
+				tmp1 = outb[top];
+				/* calculate the top output: top = top + accumulate */
+				icomplex32_add(&tmp1, &tmp2, &outb[top]);
+				/* calculate the bottom output: bottom = top - accumulate */
+				icomplex32_sub(&tmp1, &tmp2, &outb[bottom]);
+			}
+		}
+	}
+
+	/* shift back for ifft */
+	if (ifft) {
+		/*
+		 * no need to divide N as it is already done in the input side
+		 * for Q1.31 format. Instead, we need to multiply N to compensate
+		 * the shrink we did in the FFT transform.
+		 */
+		for (i = 0; i < size; i++)
+			icomplex_shift(&outb[i], len, &outb[i]);
+	}
+}
+
+/**
+ * \brief Doing Fast Fourier Transform (FFT) for mono real input buffers.
+ * \param[in] src - pointer to input buffer.
+ * \param[out] dst - pointer to output buffer, FFT output for input buffer.
+ * \param[in] size - input buffer sample count.
+ * \param[out] outb1 - pointer to output buffer 1, FFT output for input buffer 1.
+ * \param[out] outb1 - pointer to output buffer 2, FFT output for input buffer 2.
+ */
+void fft_real(struct comp_buffer *src, struct comp_buffer *dst, uint32_t size)
+{
+	struct icomplex32 *inb;
+	struct icomplex32 *outb;
+	int i;
+
+	if (src->stream.channels != 1)
+		return;
+
+	if (src->stream.size < size * sizeof(int32_t) ||
+	    dst->stream.size < size * sizeof(struct icomplex32))
+		return;
+
+	inb = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
```

Comment:
I don't think alloc & free should be used in a frequently called function. The caller should provide all the needed buffers. There could be a helper fft_init() that does such.

Reply:
It depends on how the caller wan't to use it. If the caller has comp_buffer structs and want output in this struct format also (lazy caller :-)), it can use helper here. On the other side, if the caller has intermediate buffer for holding samples, then we can use fft_internal() here (may need to remove the 'static' from the helper).

And for the alloc & free, yes, I should add helper fft_init() to do that, thanks.